etl/lineconsumer.go

```diff
@@ -0,0 +1,46 @@
+package etl
+
+import (
+	"fmt"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// LineConsumer consumes a list of lines from a channel, transforms them, and sends them to the outgoing channel.
+type LineConsumer struct {
+	wg       sync.WaitGroup
+	logger   *log.Entry
+	incoming chan string
+	outgoing chan string
+}
+
+// Start causes the LineConsumer to begin consuming lines from the incoming channel asynchronously.
+func (lc *LineConsumer) Start() {
+	go func() {
+		lc.wg.Add(1)
+		defer lc.wg.Done()
+
+		for line := range lc.incoming {
+			lc.wg.Add(1)
+			go lc.consume(line)
+		}
+	}()
+}
+
+func (lc *LineConsumer) consume(line string) {
+	lc.logger.Debug("consumed line: ", line)
+
+	lc.outgoing <- lc.transform(line)
+
+	defer lc.wg.Done()
```

Comment:
Last line of func, don't need the defer.

Defer is useful when you have multiple conditions happening in a func and you just want to ensure it returns before exiting the function - e.g.

```
func (lc *Line Consumer) consume(line string) {
  defer lc.wg.Done()
  res, err := somefuncthatreturnserrors()
  if err != nil {
    lc.logger.Error("Oh no", err.Error())
    return
  }
  // Do other things
}
```

The defer guarantees that whether your function exits because of the error or not, that defer will run. I guess it's kinda akin to Ruby's ensure or Python's finally.

Reply:
Yup. I was just doing it proactively so that if or when I refactor, it'll do the thing. Good catch though.